cylc/flow/rundb.py

```diff
@@ -916,6 +917,58 @@ def remove_columns(self, table, to_drop):
         # done
         conn.commit()
 
+    def upgrade_retry_state(self):
+        """Replace the retry state with xtriggers.
+
+        * Change *retring tasks to waiting
+        * Add the required xtrigger
+
+        Note:
+            The retry status can be safely removed as this is really a display
+            state, the retry logic revolves around the TaskActionTimer.
+
+        From:
+            cylc<8
+        To:
+            cylc>=8
+        PR:
+            #3423
+
+        Returns:
+            list - (cycle, name, status) tuples of all retrying tasks.
+
+        """
+        conn = self.connect()
+
+        for table in [self.TABLE_TASK_POOL_CHECKPOINTS, self.TABLE_TASK_POOL]:
+            tasks = list(conn.execute(
+                rf'''
+                    SELECT
+                        cycle, name, status
+                    FROM
+                        {table}
+                    WHERE
+                        status IN ('retrying', 'submit-retrying')
+                '''
+            ))
+            if tasks:
+                LOG.info(f'Upgrade retrying tasks in table {table}')
+            conn.executemany(
+                rf'''
+                    UPDATE
+                        {table}
+                    SET
+                        status='{TASK_STATUS_WAITING}'
```

Comment:
`self.TASK_STATUS_WAITING` ? The GitHub UI is not highlighting this part of the text (I see the replaced-variables in black, while the rest of string-text is in blue). Not sure if we have a test for this (last I can remember some SQL code being tested I think it was in a change by @wxtim )

Reply:
There was a test for this, I think it got lost in the merge, have pushed a commit to re-instate it.

This PR was written back when we were still aiming for `cylc restart` compatibility with Cylc7. We've given up on that one now (good idea due to major changes) so this upgrade code can probably be dropped.

Reply:
@hjoliver I think this is the last item pending before I hit approve. Should we just drop the database migration here?

Reply:
Yeah just looking at this, but I'm slightly confused. Oliver re-instated the DB test, but doesn't that test the functionality he said we can drop now??

Reply:
It seems to me @oliver-sanders reinstated the DB test before realising we can drop the upgrade code.  So I'll remove it again (and fix the conflicts).

Reply:
Cancel that, I've fixed the conflicts but will post a note to Element about upgrade from Cylc 7...